client/src/main/java/io/pravega/client/batch/impl/BatchClientImpl.java

```diff
@@ -67,53 +76,101 @@ private StreamInfo getStreamInfo(Stream stream) {
+                             return new StreamCutImpl(stream, pos);
+                         });
+    }
 
-        result.addAll(successors.getSegments());
+    private StreamSegmentsInfo getStreamSegmentInfo(final StreamCut startStreamCut, final StreamCut endStreamCut) {
+        log.debug("Start stream cut: {}, End stream cut: {}", startStreamCut, endStreamCut);
+        StreamSegmentsInfoImpl.validateStreamCuts(startStreamCut, endStreamCut);
+
+        final SortedSet<Segment> segmentSet = new TreeSet<>();
+        StreamSegmentSuccessors segments = getAndHandleExceptions(controller.getSegments(startStreamCut, endStreamCut),
+                RuntimeException::new);
+        segmentSet.addAll(segments.getSegments());
         synchronized (this) {
-            latestDelegationToken.set(successors.getDelegationToken());
+            latestDelegationToken.set(segments.getDelegationToken());
         }
-        return Iterators.transform(result.iterator(), s -> segmentToInfo(s));
+        log.debug("List of Segments between the start and end stream cuts : {}", segmentSet);
+
+        Iterator<SegmentRange> iterator = Iterators.transform(segmentSet.iterator(),
+                s -> getSegmentRange(s, startStreamCut, endStreamCut));
+        return StreamSegmentsInfoImpl.builder().segmentRangeIterator(iterator)
+                                     .startStreamCut(startStreamCut)
+                                     .endStreamCut(endStreamCut).build();
     }
 
-    private SegmentInfo segmentToInfo(Segment s) {
+    private SegmentInfo getSegmentInfo(final Segment s) {
         @Cleanup
         SegmentMetadataClient client = segmentMetadataClientFactory.createSegmentMetadataClient(s);
-
         String delegationToken;
         synchronized (this) {
             delegationToken = latestDelegationToken.get();
         }
         return client.getSegmentInfo(delegationToken);
     }
 
-    @Override
-    public <T> SegmentIterator<T> readSegment(Segment segment, Serializer<T> deserializer) {
-        @Cleanup
-        SegmentMetadataClient metadataClient = segmentMetadataClientFactory.createSegmentMetadataClient(segment);
-        SegmentInfo segmentInfo = metadataClient.getSegmentInfo(latestDelegationToken.get());
-        return new SegmentIteratorImpl<>(inputStreamFactory, segment, deserializer, segmentInfo.getStartingOffset(), segmentInfo.getWriteOffset());
-    }
-
-    @Override
-    public <T> SegmentIterator<T> readSegment(Segment segment, Serializer<T> deserializer, long startingOffset) {
-        @Cleanup
-        SegmentMetadataClient metadataClient = segmentMetadataClientFactory.createSegmentMetadataClient(segment);
-        SegmentInfo segmentInfo = metadataClient.getSegmentInfo(latestDelegationToken.get());
-        return new SegmentIteratorImpl<>(inputStreamFactory, segment, deserializer, startingOffset, segmentInfo.getWriteOffset());
+    /*
+     * Given a segment fetch its SegmentRange.
```

Comment:
`Given a segment, fetch its SegmentRange`.

Reply:
corrected it.